README.md
<h1 align="center">
  <img src="https://img.ideal-postcodes.co.uk/Ideal%20Postcodes%20Core%20Logo@3x.png" alt="Ideal Postcodes Core Interface">
</h1>

> JavaScript API for api.ideal-postcodes.co.uk

[![CI](https://github.com/ideal-postcodes/core-interface/actions/workflows/ci.yml/badge.svg)](https://github.com/ideal-postcodes/core-interface/actions/workflows/ci.yml)
![Cross Browser Testing](https://github.com/ideal-postcodes/core-interface/workflows/Cross%20Browser%20Testing/badge.svg?branch=saucelabs)
[![Release](https://github.com/ideal-postcodes/core-interface/workflows/Release/badge.svg)](https://github.com/ideal-postcodes/core-interface/actions)
[![codecov](https://codecov.io/gh/ideal-postcodes/core-interface/branch/master/graph/badge.svg)](https://codecov.io/gh/ideal-postcodes/core-interface)
![Dependency Status](https://david-dm.org/ideal-postcodes/core-interface.svg)

[![npm version](https://badge.fury.io/js/%40ideal-postcodes%2Fcore-interface.svg)](https://badge.fury.io/js/%40ideal-postcodes%2Fcore-interface)
![npm bundle size (scoped)](https://img.shields.io/bundlephobia/min/@ideal-postcodes/core-interface.svg?color=%234c1&style=popout)
![npm bundle size (scoped)](https://img.shields.io/bundlephobia/minzip/@ideal-postcodes/core-interface.svg?color=%234c1&style=popout)
[![install size](https://packagephobia.now.sh/badge?p=@ideal-postcodes/core-interface)](https://packagephobia.now.sh/result?p=@ideal-postcodes/core-interface)

`@ideal-postcodes/core-interface` is an environment agnostic implementation of the Ideal Postcodes JavaScript API client interface.

If you are looking for the browser or Node.js client which implements this interface, please check out the [downstream clients links](#downstream-clients).

## Links

- [API Documentation](https://core-interface.ideal-postcodes.dev/)
- [npm Module](https://www.npmjs.com/package/@ideal-postcodes/core-interface)
- [GitHub Repository](https://github.com/ideal-postcodes/core-interface)
- [Typings Repository](https://github.com/ideal-postcodes/openapi)
- [Fixtures Repository](https://github.com/ideal-postcodes/api-fixtures)

## Downstream Clients

- [Browser Client Repository](https://github.com/ideal-postcodes/core-browser)
- [Bundled Browser Client Repository](https://github.com/ideal-postcodes/core-browser-bundled)
- [Node.js Client Repository](https://github.com/ideal-postcodes/core-node)
- [Axios-backed Client Repository](https://github.com/ideal-postcodes/core-axios)

## Documentation

- [Usage & Configuration](#usage)
- [Quick Methods](#quick-methods)
- [Resource Methods](#resource-methods)
- [Error Handling](#error-handling)

### Methods

#### Usage

To install, pick one of the following based on your platform

```bash
# For browser client
npm install @ideal-postcodes/core-browser

# For node.js client
npm install @ideal-postcodes/core-node

# For generic interface (you need to supply your own HTTP agent)
npm install @ideal-postcodes/core-interface
```

Instantiate a client with

```javascript
import { Client } from "@ideal-postcodes/core-<package-type>";

const client = new Client({ api_key: "iddqd" });

// Only api_key is required by core-node and core-browser - all others are optional
// The agentless interface requires explicit configuration
```

[Client configuration options](https://core-interface.ideal-postcodes.dev/interfaces/client.config)

---

#### Quick Methods

The library exposes a number of simple methods to get at the most common tasks when interacting with the API

- [Lookup a Postcode](#lookup-a-postcode)
- [Search for an Address](#search-for-an-address)
- [Search for an Address by UDPRN](#search-for-an-address-by-udprn)
- [Search for an Address by UMPRN](#search-for-an-address-by-umprn)
- [Check Key Usability](#check-key-usability)

#### Lookup a Postcode

Return addresses associated with a given `postcode`

Invalid postcodes (i.e. postcode not found) return an empty array `[]`

```javascript
import { lookupPostcode } from "@ideal-postcodes/core-browser";

const postcode = "id11qd";

lookupPostcode({ client, postcode }).then(addresses => {
  console.log(addresses);
  {
    postcode: "ID1 1QD",
    line_1: "2 Barons Court Road",
    // ...etc...
  }
});
```

`lookupPostcode` [docs](https://core-interface.ideal-postcodes.dev/modules/helper_methods#lookupPostcode)

#### Search for an Address

Return addresses associated with a given `query`

```javascript
import { lookupAddress } from "@ideal-postcodes/core-browser";

const query = "10 downing street sw1a";

lookupAddress({ client, query }).then(addresses => {
  console.log(addresses);
  {
    postcode: "SW1A 2AA",
    line_1: "Prime Minister & First Lord Of The Treasury",
    // ...etc...
  }
});
```

`lookupAddress` [docs](https://core-interface.ideal-postcodes.dev/modules/helper_methods#lookupAddress)

#### Search for an Address by UDPRN

Return address for a given `udprn`

Invalid UDPRN will return `null`

```javascript
import { lookupUdprn } from "@ideal-postcodes/core-browser";

const udprn = 23747771;

lookupUdprn({ client, udprn }).then(address => {
  console.log(address);
  {
    postcode: "SW1A 2AA",
    line_1: "Prime Minister & First Lord Of The Treasury",
    // ...etc...
  }
});
```

`lookupUdprn` [docs](https://core-interface.ideal-postcodes.dev/modules/helper_methods#lookupUdprn)

#### Search for an Address by UMPRN

Return address for a given `umprn`

Invalid UMPRN will return `null`

```javascript
import { lookupUmprn } from "@ideal-postcodes/core-browser";

const umprn = 50906066;

lookupUmprn({ client, umprn }).then(address => {
  console.log(address);
  {
    postcode: "CV4 7AL",
    line_1: "Room 1, Block 1 Arthur Vick",
    // ...etc...
  }
});
```

`lookupUmprn` [docs](https://core-interface.ideal-postcodes.dev/modules/helper_methods#lookupUmprn)

#### Check Key Usability

Check if a key is currently usable

```javascript
checkKeyUsability({ client }).then((key) => {
  console.log(key.available); // => true
});
```

`checkKeyUsability` [docs](https://core-interface.ideal-postcodes.dev/modules/helper_methods#checkKeyUsability)

---

#### Resources

Resources defined in [the API documentation](https://ideal-postcodes.co.uk/documentation) are exported by the library. Each resource exposes a method (`#retrieve`, `#list`, etc) which maps to a resource action.

These methods expose a low level interface to execute HTTP requests and observe HTTP responses. They are ideal if you have a more complex query or usecase where low level access would be useful.

Resource methods return a promise with a [HTTP response object type](https://core-interface.ideal-postcodes.dev/interfaces/httpresponse.html).

#### Retrieve

Requesting a resource by ID (e.g. a postcode lookup for postcode with ID "SW1A 2AA") maps to the `#retrieve` method.

The first argument is the client object. The second is the resource ID. The last argument is an object which accepts `header` and `query` attributes that map to HTTP header and the request querystring.

```javascript
resourceName.retrieve(client, "id", {
  query: {
    api_key: "foo",
    tags: "this,that,those",
    licensee: "sk_99dj3",
  },
  header: {
    "IDPC-Source-IP": "8.8.8.8",
  },
  timeout: 5000,
});
```

#### List

Requesting a resource endpoint (e.g. an address query to `/addresses`) maps to the `#list` method.

```javascript
resourceName.list(client, {
  query: {
    api_key: "foo",
    query: "10 downing street",
  },
  header: {
    "IDPC-Source-IP": "8.8.8.8",
  },
  timeout: 5000,
});
```

The first argument is the client. The second is an object which accepts `header` and `query` attributes that map to HTTP header and the request querystring.

#### Custom Actions

Some endpoints are defined as custom actions, e.g. `/keys/:key/usage`. These can be invoked using the name of the custom action.

E.g. for [key usage data extraction](https://ideal-postcodes.co.uk/documentation/keys#usage)

```javascript
keys.usage(client, api_key, {
  query: {
    tags: "checkout,production",
  },
  header: {
    Authorization: 'IDEALPOSTCODES user_token="foo"',
  },
  timeout: 5000,
});
```

---

#### Resource Methods

Listed below are the available resources exposed by the library:

- [Postcodes](#postcodes)
- [Addresses](#addresses)
- [Autocomplete](#autocomplete)
- [UDPRN](#udprn)
- [UMPRN](#umprn)
- [Keys](#keys)

#### Postcodes

Retrieve addresses for a postcode.

```javascript
import { postcodes } from "@ideal-postcodes/core-browser";
postcodes
  .retrieve(client, "SW1A2AA", {
    header: {
      Authorization: 'IDEALPOSTCODES api_key="iddqd"',
    },
  })
  .then((response) => {
    const addresses = response.body.result;
  })
  .catch((error) => logger(error));
```

[Postcode resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/postcodes)

[Postcode resource docs](https://core-interface.ideal-postcodes.dev/interfaces/postcoderesource.html)

#### Addresses

Search for an address

```javascript
import { addresses } from "@ideal-postcodes/core-browser";

addresses
  .list(client, {
    query: {
      query: "10 Downing street",
    },
    header: {
      Authorization: 'IDEALPOSTCODES api_key="iddqd"',
    },
  })
  .then((response) => {
    const addresses = response.body.result.hits;
  })
  .catch((error) => logger(error));
```

[Address resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/addresses)

[Address resource client docs](https://core-interface.ideal-postcodes.dev/modules/resources_addresses.html)

#### Autocomplete

Autocomplete an address given an address partial

```javascript
import { autocomplete } from "@ideal-postcodes/core-browser";

autocomplete
  .list(client, {
    query: {
      query: "10 Downing stre",
    },
    header: {
      Authorization: 'IDEALPOSTCODES api_key="iddqd"',
    },
  })
  .then((response) => {
    const suggestions = response.body.result.hits;
  })
  .catch((error) => logger(error));
```

[Autocomplete resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/autocomplete)

[Autocomplete resource client docs](https://core-interface.ideal-postcodes.dev/modules/resources_autocomplete.html)

#### UDPRN

Retrieve an address given a UDPRN

```javascript
import { udprn } from "@ideal-postcodes/core-browser";

udprn
  .retrieve(client, "12345678", {
    header: {
      Authorization: 'IDEALPOSTCODES api_key="iddqd"',
    },
  })
  .then((response) => {
    const address = response.body.result;
  })
  .catch((error) => logger(error));
```

[UDPRN resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/udprn)

[UDPRN resource client docs](https://core-interface.ideal-postcodes.dev/modules/resources_udprn.html)

#### UMPRN

Retrieve a multiple residence premise given a UMPRN

```javascript
import { umprn } from "@ideal-postcodes/core-browser";

umprn
  .retrieve(client, "87654321", {
    header: {
      Authorization: 'IDEALPOSTCODES api_key="iddqd"',
    },
  })
  .then((response) => {
    const address = response.body.result;
  })
  .catch((error) => logger(error));
```

[UMPRN resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/umprn)

[UMPRN resource client docs](https://core-interface.ideal-postcodes.dev/modules/resources_umprn.html)

#### Keys

Find out if a key is available

```javascript
import { keys } from "@ideal-postcodes/core-browser";

keys
  .retrieve(client, "iddqd", {})
  .then((response) => {
    const { available } = response.body.result;
  })
  .catch((error) => logger(error));
```

[Method docs](https://core-interface.ideal-postcodes.dev/modules/resources_keys.html#retrieve)

Get private information on key (requires user_token)

```javascript
import { keys } from "@ideal-postcodes/core-browser";

keys
  .retrieve(client, "iddqd", {
    header: {
      Authorization: 'IDEALPOSTCODES user_token="secret-token"',
    },
  })
  .then((response) => {
    const key = response.body.result;
  })
  .catch((error) => logger(error));
```

[Method docs](https://core-interface.ideal-postcodes.dev/modules/resources_keys.html#retrieve)

Get key usage data

```javascript
import { keys } from "@ideal-postcodes/core-browser";

keys
  .usage(client, "iddqd", {
    header: {
      Authorization: 'IDEALPOSTCODES user_token="secret-token"',
    },
  })
  .then((response) => {
    const key = response.body.result;
  })
  .catch((error) => logger(error));
```

[Method docs](https://core-interface.ideal-postcodes.dev/modules/resources_keys.html#usage)

[Keys resource HTTP API documentation](https://ideal-postcodes.co.uk/documentation/keys)

[Key resource client docs](https://core-interface.ideal-postcodes.dev/modules/resources_keys.html)

---

#### Error Handling

This library exports a static variable `errors` which contains custom error constructors that wrap specific API errors. These constructors can be used to test for specific cases using the `instanceof` operator.

For example:

```javascript
const { IdpcInvalidKeyError } = Client.errors;

try {
  const addresses = await lookupPostcode({ client, postcode: "SW1A2AA" });
} catch (error) {
  if (error instanceof IdpcInvalidKeyError) {
    // Handle an invalid key error
  }
}
```

Not all specific API errors will be caught. If a specific API error does not have an error constructor defined, a more generic error (determined by the HTTP status code) will be returned.

For example:

```javascript
import {
  IdpcRequestFailedError,
  lookupPostcode,
} from "@ideal-postcodes/core-browser";

try {
  const addresses = await lookupPostcode({ client, postcode: "SW1A2AA" });
} catch (error) {
  if (error instanceof IdpcRequestFailedError) {
    // IdpcRequestFailedError indicates a 402 response code
    // Possibly the key balance has been depleted
  }
}
```

You may view a [sketch of the error prototype chain](#error-prototype-chain).

---

### Core Interface Errors

For more advanced use cases, this core-interface library provides:

- Class implementations for [Ideal Postcodes API errors](https://core-interface.ideal-postcodes.dev/classes/errors.idpcapierror) that inherit from `Error`
- A [parser](https://core-interface.ideal-postcodes.dev/modules/errors#parse) that converts raw error data into one of these error instances

#### Error Usage

Aside from inspecting the HTTP request status code and/or JSON body response codes, you may also test for specific error instances.

Errors that don't inherit from [`IdealPostcodesError`](https://core-interface.ideal-postcodes.dev/classes/errors.idealpostcodeserror) would indicate some kind of error external to the API (e.g. bad network, request timeout).

```javascript
import { errors } from "@ideal-postcodes/core-browser";
const { IdpcPostcodeNotFoundError } = errors;

// Handle a specific error
if (error instanceof IdpcPostcodeNotFoundError) {
  // You got yourself an invalid API Key
}

// Alternatively use a switch statement
switch (true) {
  case error instanceof IdpcPostcodeNotFoundError:
  // You got yourself an invalid API Key
  default:
  // Default error handling path
}
```

#### Error Prototype Chain

All errors inherit from JavaScript's `Error` prototype.

Errors are grouped by HTTP status code classes.

Specific errors may be supplied for the following reasons:

- Convenience. They are frequently tested for (e.g. invalid postcode, postcode not found)
- Useful for debug purposes during the implementation stages

```javascript
Prototype Chain

# Parent class inherits from Javascript Error. Returned if no JSON Response body
IdealPostcodesError < Error
|
|- IdpcApiError # Generic Error Class, returned if JSON response body present
   |
   |- IdpcBadRequestError          # 400 Errors
   |- IdpcUnauthorisedError        # 401 Errors
   |- IdpcRequestFailedError       # 402 Errors
   |  |- IdpcBalanceDepletedError
   |  |- IdpcLimitReachedError
   |
   |- IdpcResourceNotFoundError    # 404 Errors
   |  |- IdpcPostcodeNotFoundError
   |  |- IdpcKeyNotFoundError
   |  |- IdpcUdprnNotFoundError
   |  |- IdpcUmprnNotFoundError
   |
   |- IdpcServerError              # 500 Errors
```

#### Error Parser

The error parser consumes a HTTP API response and returns the correct error instance.

```javascript
import { errors } from "@ideal-postcodes/core-browser";
const { parse, IdpcPostcodeNotFoundError } = errors;

const invalidPostcodeUrl = "https://api.ideal-postcodes.co.uk/v1/postcodes/bad_postcode?api_key=iddqd"

const response = await fetch(invalidPostcodeUrl);

// Generate an error object if present, otherwise returns `undefined`
const error = parse(response);

// Handle the error
if (error instanceof IdpcPostcodeNotFoundError) {...}
```

## Test

```bash
npm test
```

## Licence

MIT


CHANGELOG.md
## [3.2.1](https://github.com/ideal-postcodes/core-interface/compare/3.2.0...3.2.1) (2022-08-31)


### Bug Fixes

* **Internal:** Export internal address lookup methods ([4965c73](https://github.com/ideal-postcodes/core-interface/commit/4965c739380e8a334ed72e9122c16f9b48413fbd))

# [3.2.0](https://github.com/ideal-postcodes/core-interface/compare/3.1.0...3.2.0) (2022-08-05)


### Features

* **OpenAPI:** Update to 3.1.0 ([7f5c80b](https://github.com/ideal-postcodes/core-interface/commit/7f5c80bd9b98f70b44505dfde6236c1298205778))

# [3.1.0](https://github.com/ideal-postcodes/core-interface/compare/3.0.0...3.1.0) (2022-04-04)


### Features

* **USA:** Resolve addreses in US format ([76aff0c](https://github.com/ideal-postcodes/core-interface/commit/76aff0c9c462c7e53f890d04b48bdd764eab6acd))

# [3.0.0](https://github.com/ideal-postcodes/core-interface/compare/2.0.3...3.0.0) (2022-02-14)


### Features

* **OpenAPI:** Migrate from api-typings to openapi spec ([62c41b2](https://github.com/ideal-postcodes/core-interface/commit/62c41b24a3064f68a5c7216fdc9953ca71d68c27))
* **Resolve:** Resolve address suggestions ([3c1ad61](https://github.com/ideal-postcodes/core-interface/commit/3c1ad612787ec54a4066c09763aa77f11c0a7da1))


### BREAKING CHANGES

* **OpenAPI:** Core-Interface typings now uses
@ideal-postcodes/openapi

# [3.0.0-beta.1](https://github.com/ideal-postcodes/core-interface/compare/2.0.3...3.0.0-beta.1) (2022-02-03)


### Features

* **OpenAPI:** Migrate from api-typings to openapi spec ([06ce31d](https://github.com/ideal-postcodes/core-interface/commit/06ce31d6de53b65837fbb30ba44f88d343c51c90))
* **Resolve:** Resolve address suggestions ([c2cfef7](https://github.com/ideal-postcodes/core-interface/commit/c2cfef7ee4fcc5ec8a66989278f4ab5dbf778723))


### BREAKING CHANGES

* **OpenAPI:** Core-Interface typings now uses
@ideal-postcodes/openapi

## [2.0.3](https://github.com/ideal-postcodes/core-interface/compare/2.0.2...2.0.3) (2021-07-23)


### Bug Fixes

* **ESM:** Pin target output to ES2020 ([78a8b74](https://github.com/ideal-postcodes/core-interface/commit/78a8b7464290710b69bb9d9abbff5413cfc7dca6))

## [2.0.2](https://github.com/ideal-postcodes/core-interface/compare/2.0.1...2.0.2) (2021-07-19)


### Bug Fixes

* **Exports:** Drop namespaced exports ([9717306](https://github.com/ideal-postcodes/core-interface/commit/9717306545d6f61cd15ad7ea313727b3f3fd0446))
* **Typings:** Move into dependencies ([8caef68](https://github.com/ideal-postcodes/core-interface/commit/8caef684af0fbfdcc705efc541625d1b55ac3f21))

## [2.0.1](https://github.com/ideal-postcodes/core-interface/compare/2.0.0...2.0.1) (2021-06-04)


### Bug Fixes

* **Client Typings:** Make all but api_key config optional ([716cba1](https://github.com/ideal-postcodes/core-interface/commit/716cba155e4091450abfa82129ac938d61a32edd))

# [2.0.0](https://github.com/ideal-postcodes/core-interface/compare/1.9.0...2.0.0) (2021-06-04)


### Code Refactoring

* **Defaults:** Export `defaults` object ([6b9b698](https://github.com/ideal-postcodes/core-interface/commit/6b9b6981abd517061621436a2afef0f4f62cddb7))


### Features

* **Version 2:** Reduce package size ([cfcae8a](https://github.com/ideal-postcodes/core-interface/commit/cfcae8a7087708820ec0ca1b2d97df3dabd056f5))


### BREAKING CHANGES

* **Version 2:** - Package now exports a `defaults` object
- Client.defaults has been removed
- All client config is now stored in `client.config`
- All resources have been removed from the client. Instead retrieve
these from the library and inject the client. E.g.
`client.postcodes.retrieve` becomes `postcodes.retrieve(client, ...)`
- Helper methods (like lookupPostcode, ping) have been removed from the client.
Instead retrieve these from teh library and inject the client. E.g.
`client.lookupPostcode` becomes `lookupPostcode(client, ...)`
* **Defaults:** Uppercased default variables no longer exported.
Instead `defaults` object is supplied

# [1.9.0](https://github.com/ideal-postcodes/core-interface/compare/1.8.1...1.9.0) (2021-01-15)


### Features

* **Tagging:** Allow tags to be set by client ([5f40c0f](https://github.com/ideal-postcodes/core-interface/commit/5f40c0f98c8a19b46525d0c29aa9e37a339cf747))

## [1.8.1](https://github.com/ideal-postcodes/core-interface/compare/1.8.0...1.8.1) (2020-11-27)


### Bug Fixes

* **tsconfig:** Bump tsconfig ([37504cd](https://github.com/ideal-postcodes/core-interface/commit/37504cdc2e4ecf6a610d672bce60cfdbb61a2e86))

# [1.8.0](https://github.com/ideal-postcodes/core-interface/compare/1.7.0...1.8.0) (2020-10-22)


### Features

* **ESM:** Target ESNext in ESM builds ([d44f960](https://github.com/ideal-postcodes/core-interface/commit/d44f960b314a89881b6afd67f9c8f44cc0a7621a))

# [1.7.0](https://github.com/ideal-postcodes/core-interface/compare/1.6.0...1.7.0) (2020-10-21)


### Features

* **ESM:** Publish ESM build ([acf45bb](https://github.com/ideal-postcodes/core-interface/commit/acf45bbac2f4f0b3872115289e1d35071991abae))

# [1.6.0](https://github.com/ideal-postcodes/core-interface/compare/1.5.0...1.6.0) (2020-07-27)


### Features

* **Query:** Allow non-string query attributes ([2de56cb](https://github.com/ideal-postcodes/core-interface/commit/2de56cb1905e04b6376e5a2a4dc52a9b65aff7be))

# [1.5.0](https://github.com/ideal-postcodes/core-interface/compare/1.4.0...1.5.0) (2020-02-05)


### Features

* **Keys:** Add ability to check licensee ([46d1cd6](https://github.com/ideal-postcodes/core-interface/commit/46d1cd63894d37f727ade827cb1a273fab3e3cf8))

# [1.4.0](https://github.com/ideal-postcodes/core-interface/compare/1.3.0...1.4.0) (2019-12-06)


### Bug Fixes

* **Browser Tests:** Migrate to SL ([2b96ae5](https://github.com/ideal-postcodes/core-interface/commit/2b96ae5))
* **README:** Check against Markdown lint rules and correct grammar ([b210165](https://github.com/ideal-postcodes/core-interface/commit/b210165))


### Features


lib/error.ts
/**
 * @module Errors
 *
 * @description Exports error classes which may be returned by this client
 */

import { HttpResponse } from "./agent";
import { components } from "@ideal-postcodes/openapi";

export type ApiErrorResponse = components["schemas"]["ErrorResponse"];

/**
 * IdealPostcodesErrorOptions
 *
 * IdealPostcodesError requires only a HTTP status from a failed API request
 */
interface IdealPostcodesErrorOptions {
  message: string;
  httpStatus: number;
  metadata?: Metadata;
}

/**
 * Metadata
 *
 * An abitrary object which stores metadata specific to the concrete client
 * implementation
 */
interface Metadata {
  [k: string]: unknown;
}

// Take note of https://github.com/Microsoft/TypeScript/issues/13965

/**
 * IdealPostcodesError
 *
 * Base error class for all API responses that return an error. This class
 * is used where a JSON body is not provided or invalid
 * E.g. 503 rate limit response, JSON parse failure response
 */
export class IdealPostcodesError extends Error {
  __proto__: Error;

  public httpStatus: number;
  public metadata: Metadata;

  /**
   * Instantiate IdealPostcodesError
   */
  constructor(options: IdealPostcodesErrorOptions) {
    const trueProto = new.target.prototype;
    super();
    this.__proto__ = trueProto;

    const { message, httpStatus, metadata = {} } = options;
    this.message = message;
    this.name = "Ideal Postcodes Error";
    this.httpStatus = httpStatus;
    this.metadata = metadata;

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, IdealPostcodesError);
    }
  }
}

/**
 * IdpcApiError
 *
 * Base error class for API responses with a JSON body. Typically a subclass
 * will be used to capture the error category (e.g. 400, 401, 500, etc)
 */
export class IdpcApiError extends IdealPostcodesError {
  /**
   * Raw HTTP response
   */
  public response: HttpResponse;

  /**
   * Returns an API error instance
   */
  constructor(httpResponse: HttpResponse) {
    super({
      httpStatus: httpResponse.httpStatus,
      message: httpResponse.body.message,
    });
    this.response = httpResponse;
  }
}

/**
 * IdpcBadRequestError
 *
 * Captures API responses that return a 400 (Bad Request Error) response
 *
 * Examples include:
 * - Invalid syntax submitted
 * - Invalid date range submitted
 * - Invalid tag submitted
 */
export class IdpcBadRequestError extends IdpcApiError {}

/**
 * IdpcUnauthorisedError
 *
 * Captures API responses that return a 401 (Unauthorised) response
 *
 * Examples include:
 * - Invalid api_key
 * - Invalid user_token
 * - Invalid licensee
 */
export class IdpcUnauthorisedError extends IdpcApiError {}

/**
 * IpdcInvalidKeyError
 *
 * Invalid API Key presented for request
 */
export class IdpcInvalidKeyError extends IdpcUnauthorisedError {}

/**
 * IdpcRequestFailedError
 *
 * Captures API responses that return a 402 (Request Failed) response
 *
 * Examples include:
 * - Key balance depleted
 * - Daily key limit reached
 */
export class IdpcRequestFailedError extends IdpcApiError {}

/**
 * IdpcBalanceDepleted
 *
 * Balance on key has been depleted
 */
export class IdpcBalanceDepletedError extends IdpcRequestFailedError {}

/**
 * IdpcLimitReachedError
 *
 * Limit reached. One of your lookup limits has been breached for today. This
 * could either be your total daily limit on your key or the individual IP
 * limit. You can either wait for for the limit to reset (after a day) or
 * manually disable or increase your limit.
 */
export class IdpcLimitReachedError extends IdpcRequestFailedError {}

/**
 * IdpcResourceNotFoundError
 *
 * Captures API responses that return a 404 (Resource Not Found) response
 *
 * Examples include:
 * - Postcode not found
 * - UDPRN not found
 * - Key not found
 */
export class IdpcResourceNotFoundError extends IdpcApiError {}

/**
 * IdpcPostcodeNotFoundError
 *
 * Requested postcode does not exist
 */
export class IdpcPostcodeNotFoundError extends IdpcResourceNotFoundError {}

/**
 * IdpcKeyNotFoundError
 *
 * Requested API Key does not exist
 */
export class IdpcKeyNotFoundError extends IdpcResourceNotFoundError {}

/**
 * IdpcUdprnNotFoundError
 *
 * Requested UDPRN does not exist
 */
export class IdpcUdprnNotFoundError extends IdpcResourceNotFoundError {}

/**
 * IdpcUmprnNotFoundError
 *
 * Requested UMPRN does not exist
 */
export class IdpcUmprnNotFoundError extends IdpcResourceNotFoundError {}

/**
 * IdpcServerError
 *
 * Captures API responses that return a 500 (Server Error) response
 */
export class IdpcServerError extends IdpcApiError {}

// 200 Responses
const OK = 200;

// 300 Responses
const REDIRECT = 300;

// 400 Responses
const BAD_REQUEST = 400;

// 401 Responses
const UNAUTHORISED = 401;
const INVALID_KEY = 4010;

// 402 Responses
const PAYMENT_REQUIRED = 402;
const BALANCE_DEPLETED = 4020;
const LIMIT_REACHED = 4021;

// 404 Responses
const NOT_FOUND = 404;
const POSTCODE_NOT_FOUND = 4040;
const KEY_NOT_FOUND = 4042;
const UDPRN_NOT_FOUND = 4044;
const UMPRN_NOT_FOUND = 4046;

// 500 Responses
const SERVER_ERROR = 500;

const isSuccess = (code: number): boolean => {
  if (code < OK) return false;
  if (code >= REDIRECT) return false;
  return true;
};

const isObject = (o: unknown): o is object => {
  if (o === null) return false;
  if (typeof o !== "object") return false;
  return true;
};

const isErrorResponse = (body: unknown): body is ApiErrorResponse => {
  if (!isObject(body)) return false;
  if (typeof (body as ApiErrorResponse).message !== "string") return false;
  if (typeof (body as ApiErrorResponse).code !== "number") return false;
  return true;
};

/**
 * parse
 *
 * Parses API responses and returns an error for non 2xx responses
 *
 * Upon detecting an error an instance of IdealPostcodesError is returned
 */
export const parse = (response: HttpResponse): Error | void => {
  const { httpStatus, body } = response;

  if (isSuccess(httpStatus)) return;

  if (isErrorResponse(body)) {
    // Test for specific API errors of interest
    const { code } = body;
    if (code === INVALID_KEY) return new IdpcInvalidKeyError(response);
    if (code === POSTCODE_NOT_FOUND)
      return new IdpcPostcodeNotFoundError(response);
    if (code === KEY_NOT_FOUND) return new IdpcKeyNotFoundError(response);
    if (code === UDPRN_NOT_FOUND) return new IdpcUdprnNotFoundError(response);
    if (code === UMPRN_NOT_FOUND) return new IdpcUmprnNotFoundError(response);
    if (code === BALANCE_DEPLETED)
      return new IdpcBalanceDepletedError(response);
    if (code === LIMIT_REACHED) return new IdpcLimitReachedError(response);

    // If no API errors of interest detected, fall back to http status code
    if (httpStatus === NOT_FOUND)
      return new IdpcResourceNotFoundError(response);
    if (httpStatus === BAD_REQUEST) return new IdpcBadRequestError(response);
    if (httpStatus === PAYMENT_REQUIRED)
      return new IdpcRequestFailedError(response);
    if (httpStatus === UNAUTHORISED) return new IdpcUnauthorisedError(response);
    if (httpStatus === SERVER_ERROR) return new IdpcServerError(response);
  }

  // Generate generic error (backstop)
  return new IdealPostcodesError({ httpStatus, message: JSON.stringify(body) });
};
